elasticsearch/src/test/java/org/hibernate/search/elasticsearch/test/ElasticsearchFlushIT.java

```diff
@@ -0,0 +1,145 @@
+/*
+ * Hibernate Search, full-text search for your domain model
+ *
+ * License: GNU Lesser General Public License (LGPL), version 2.1 or later
+ * See the lgpl.txt file in the root directory or <http://www.gnu.org/licenses/lgpl-2.1.html>.
+ */
+package org.hibernate.search.elasticsearch.test;
+
+import java.io.Serializable;
+import java.util.Collections;
+import java.util.Locale;
+
+import org.hibernate.search.annotations.Field;
+import org.hibernate.search.annotations.Indexed;
+import org.hibernate.search.backend.LuceneWork;
+import org.hibernate.search.bridge.util.impl.ContextualExceptionBridgeHelper;
+import org.hibernate.search.cfg.Environment;
+import org.hibernate.search.elasticsearch.cfg.ElasticsearchEnvironment;
+import org.hibernate.search.elasticsearch.cfg.IndexSchemaManagementStrategy;
+import org.hibernate.search.elasticsearch.client.impl.JestClient;
+import org.hibernate.search.engine.integration.impl.ExtendedSearchIntegrator;
+import org.hibernate.search.engine.service.spi.ServiceManager;
+import org.hibernate.search.engine.service.spi.ServiceReference;
+import org.hibernate.search.engine.spi.DocumentBuilderIndexedEntity;
+import org.hibernate.search.engine.spi.EntityIndexBinding;
+import org.hibernate.search.query.engine.spi.HSQuery;
+import org.hibernate.search.spi.DefaultInstanceInitializer;
+import org.hibernate.search.testsupport.TestForIssue;
+import org.hibernate.search.testsupport.junit.SearchFactoryHolder;
+
+import org.junit.Assert;
+import org.junit.Rule;
+import org.junit.Test;
+
+import io.searchbox.indices.settings.UpdateSettings;
+import org.apache.lucene.search.MatchAllDocsQuery;
+
+public class ElasticsearchFlushIT {
+
+	@Rule
+	public SearchFactoryHolder sfHolder = new SearchFactoryHolder( Entity1.class, Entity2.class )
+			.withProperty( "hibernate.search.default." + Environment.INDEX_MANAGER_IMPL_NAME, "elasticsearch" )
+			.withProperty( "hibernate.search.default." + ElasticsearchEnvironment.REFRESH_AFTER_WRITE, "false" )
+			.withProperty(
+					"hibernate.search.default." + ElasticsearchEnvironment.INDEX_SCHEMA_MANAGEMENT_STRATEGY,
+					IndexSchemaManagementStrategy.RECREATE_DELETE.name()
+			)
+			.withIdProvidedImplicit( true );
+
+
+	@Test
+	@TestForIssue(jiraKey = "HSEARCH-2491")
+	public void testFlushByEntity() throws Exception {
+
+		increaseRefreshTime( Entity1.class, Entity2.class );
+
+		Entity1 entity1 = new Entity1( 1 );
+		Entity2 entity2 = new Entity2( 1 );
+
+		indexAsStream( "E1:1", entity1 );
+		indexAsStream( "E2:1", entity2 );
+
+		assertIndexed( Entity1.class, 0 );
+		assertIndexed( Entity2.class, 0 );
+
+		flush( Entity1.class );
```

Comment:
I think there's a race here: the _indexAsStream_ operations are merely scheduling work into a queue so that they can be batched with periodic writes. The _flush_ operation is issuing a flush request on the ES node, but it's not flushing this queue first.

I'll merge it anyway as it doesn't seem to affect this test in practice, and there's good progress here..